return Box<dyn Reply>

due to Reply trait consuming self you can't return Box<dyn Reply> from an `and_then` handler and allow for multiple Reply types eg:
```Rust
#![deny(warnings)]
extern crate futures;
extern crate warp;

use warp::Filter;

fn reply() -> impl futures::Future<Item = Box<dyn warp::Reply + 'static>, Error = warp::Rejection> {
    futures::future::ok(Box::new("hello world") as Box<dyn warp::Reply>)
}

fn main() {
    // Match any request and return hello world!
    let routes = warp::any().and_then(reply);

    warp::serve(routes).run(([127, 0, 0, 1], 3030));
}
```
does not compile:
```
error[E0277]: the trait bound `std::boxed::Box<(dyn warp::Reply + 'static)>: warp::Reply` is not satisfied
  --> examples/hello.rs:15:5
   |
15 |     warp::serve(routes).run(([127, 0, 0, 1], 3030));
   |     ^^^^^^^^^^^ the trait `warp::Reply` is not implemented for `std::boxed::Box<(dyn warp::Reply + 'static)>`
   |
   = note: required because of the requirements on the impl of `warp::Reply` for `(std::boxed::Box<(dyn warp::Reply + 'static)>,)`
   = note: required because of the requirements on the impl of `warp::server::IntoWarpService` for `warp::filter::and_then::AndThen<impl warp::Filter+std::marker::Copy, fn() -> impl futures::Future {reply}>`
   = note: required by `warp::serve`

error[E0599]: no method named `run` found for type `warp::Server<warp::filter::and_then::AndThen<impl warp::Filter+std::marker::Copy, fn() -> impl futures::Future {reply}>>` in the current scope
  --> examples/hello.rs:15:25
   |
15 |     warp::serve(routes).run(([127, 0, 0, 1], 3030));
   |                         ^^^
   |
   = note: the method `run` exists but the following trait bounds were not satisfied:
           `(std::boxed::Box<(dyn warp::Reply + 'static)>,) : warp::Reply`
           `warp::filter::and_then::AndThen<impl warp::Filter+std::marker::Copy, fn() -> impl futures::Future {reply}> : warp::server::IntoWarpService`
           `warp::filter::service::FilteredService<warp::filter::and_then::AndThen<impl warp::Filter+std::marker::Copy, fn() -> impl futures::Future {reply}>> : warp::server::WarpService`

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: Could not compile `warp`.

To learn more, run the command again with --verbose.
```

fiddling around yesterday, we were able to overcome the issue with the following solution

```Rust
trait BoxedReply: Reply {
    fn into_boxed_responseself: Box<Self>) -> Response;
}

impl<T> BoxedReply for T where T: Reply {
    fn into_boxed_response(self: Box<Self>) -> Response {
        (*self).into_response()
    }
}

/// wrapper type to allow returning Boxed Reply's
#[allow(missing_debug_implementations)]
pub struct DynReply(Box<dyn BoxedReply>);

impl Reply for DynReply {
    fn into_response(self) -> Response {
        self.0.into_boxed_response()
    }
}

///return a struct that has an inner Dyn Reply and implements Reply
pub fn boxed<T: Reply + 'static>(t: T) -> DynReply {
    DynReply(Box::new(t))
}
```
is there a better solution?
cc @brunoczim

--- a/examples/dyn_reply.rs
+++ b/examples/dyn_reply.rs
@@ -1,0 +1,18 @@
+#![deny(warnings)]
+use warp::{http::StatusCode, Filter};
+
+async fn dyn_reply(word: String) -> Result<Box<dyn warp::Reply>, warp::Rejection> {
+    if &word == "hello" {
+        // a cast is needed for now, see https://github.com/rust-lang/rust/issues/60424
+        Ok(Box::new("world") as Box<dyn warp::Reply>)
+    } else {
+       Ok(Box::new(StatusCode::BAD_REQUEST) as Box<dyn warp::Reply>)
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let routes = warp::path::param().and_then(dyn_reply);
+
+    warp::serve(routes).run(([127, 0, 0, 1], 3030)).await;
+}

--- a/src/reply.rs
+++ b/src/reply.rs
@@ -47,6 +47,7 @@ use serde_json;
 use crate::reject::IsReject;
 // This re-export just looks weird in docs...
 pub(crate) use self::sealed::Reply_;
+use self::sealed::{BoxedReply, Internal};
 #[doc(hidden)]
 pub use crate::filters::reply as with;
 
@@ -226,8 +227,7 @@ where
 ///
 /// let route = warp::any().map(handler);
 /// ```
-//NOTE: This list is duplicated in the module documentation.
-pub trait Reply: Send {
+pub trait Reply: BoxedReply + Send {
     /// Converts the given value into a [`Response`].
     ///
     /// [`Response`]: type.Response.html
@@ -290,6 +290,12 @@ pub trait Reply: Send {
         }
     }
     */
+}
+
+impl Reply for Box<dyn Reply> {
+    fn into_response(self) -> Response {
+        self.boxed_into_response(Internal)
+    }
 }
 
 fn _assert_object_safe() {
@@ -540,6 +546,23 @@ mod sealed {
             self.0
         }
     }
+
+    #[allow(missing_debug_implementations)]
+    pub struct Internal;
+
+    // Implemented for all types that implement `Reply`.
+    //
+    // A user doesn't need to worry about this, it's just trait
+    // hackery to get `Box<dyn Reply>` working.
+    pub trait BoxedReply {
+        fn boxed_into_response(self: Box<Self>, internal: Internal) -> Response;
+    }
+
+    impl<T: Reply> BoxedReply for T {
+        fn boxed_into_response(self: Box<Self>, _: Internal) -> Response {
+            self.into_response()
+        }
+    }
 }
 
 #[cfg(test)]
@@ -568,4 +591,10 @@ mod tests {
         assert_eq!(res.status(), 500);
     }
 
+    #[test]
+    fn boxed_reply() {
+        let r: Box<dyn Reply> = Box::new(reply());
+        let resp = r.into_response();
+        assert_eq!(resp.status(), 200);
+    }
 }

